src/ngraph/attribute_visitor.hpp

```diff
@@ -0,0 +1,46 @@
+//*****************************************************************************
+// Copyright 2017-2019 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#pragma once
+
+#include <string>
+
+#include "ngraph/partial_shape.hpp"
+#include "ngraph/type/element_type.hpp"
+
+namespace ngraph
+{
+    class VisitorAdapter;
+
+    /// \brief Visits the attributes of a node.
+    ///
+    /// Attributes are the values set when building a graph which are not
+    /// computed as the graph executes. Values computed from the graph topology and attributes
+    /// during compilation are not attributes.
+    class AttributeVisitor
+    {
+    public:
+        virtual ~AttributeVisitor() {}
+        virtual void on_attribute(const std::string& name, std::string& value) = 0;
+        virtual void on_attribute(const std::string& name, element::Type& value) = 0;
```

Comment:
Why do you add `on_attribute` methods for `element::Type` and `PartialShape`? It is not basic types.

Why don't you want to use `VisitorAdapter` for all types? It looks like current solution has some specific adapters which are implemented in `on_attribute` methods (for `std::string`, `element::Type`, `bool` and etc.).

We can divide access to attribute and it serialisation/deserialisation, if visitor contains only one method `on_attribute` with  `VisitorAdapter` argument and we have default adapters for all basic types.

Reply:
Mainly because I'm hooking into the current serializer right now and those two are serialized in ad-hoc ways. If we made switched to a new uniform format, we wouldn't need those special-cased.

Reply:
Ok, I got it.

But why cannot you remove these special-cased methods and add a code on the serialiser part?
I understand that it can be WA but it will allow us to avoid some redundant methods in the new API.
Because users can be confused why these objects have privileges over other objects and can be used without adapters.